File: ios/Runner/NativeAdFactory.swift
import google_mobile_ads

class NativeAdFactory: NSObject, FLTNativeAdFactory {
    func createNativeAd(_ nativeAd: GADNativeAd,
                        customOptions: [AnyHashable : Any]? = nil) -> GADNativeAdView? {
        let nibView = Bundle.main.loadNibNamed("NativeAdView", owner: nil, options: nil)!.first
        let nativeAdView = nibView as! GADNativeAdView

        // Set the headline
        if let headlineView = nativeAdView.headlineView as? UILabel {
            headlineView.text = nativeAd.headline
        }

        // Set the body text
        if let bodyView = nativeAdView.bodyView as? UILabel {
            bodyView.text = nativeAd.body
        }

        // Set the media view
        if let mediaView = nativeAdView.mediaView {
            mediaView.mediaContent = nativeAd.mediaContent
        }

        // Set the icon
        if let iconView = nativeAdView.iconView as? UIImageView {
            iconView.image = nativeAd.icon?.image
        }

        // Set call to action
        if let callToActionView = nativeAdView.callToActionView as? UIButton {
            callToActionView.setTitle(nativeAd.callToAction, for: .normal)
        }

        nativeAdView.nativeAd = nativeAd
        return nativeAdView
    }
}

File: android/app/src/main/kotlin/com/thesis/generator/ai/NativeAdFactory.kt
package com.thesis.generator.ai

import android.content.Context
import android.view.LayoutInflater
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import com.google.android.gms.ads.nativead.NativeAd
import com.google.android.gms.ads.nativead.NativeAdView
import io.flutter.plugins.googlemobileads.GoogleMobileAdsPlugin

class NativeAdFactory(private val context: Context) : GoogleMobileAdsPlugin.NativeAdFactory {
    override fun createNativeAd(
        nativeAd: NativeAd,
        customOptions: MutableMap<String, Any>?
    ): NativeAdView {
        val nativeAdView = LayoutInflater.from(context)
            .inflate(R.layout.native_ad_layout, null) as NativeAdView

        with(nativeAdView) {
            // Headline
            nativeAd.headline?.let {
                findViewById<TextView>(R.id.ad_headline).text = it
                headlineView = findViewById(R.id.ad_headline)
            }

            // Body
            nativeAd.body?.let {
                findViewById<TextView>(R.id.ad_body).text = it
                bodyView = findViewById(R.id.ad_body)
            }

            // Media
            mediaView = findViewById(R.id.ad_media)

            // App Icon
            nativeAd.icon?.let {
                findViewById<ImageView>(R.id.ad_app_icon).setImageDrawable(it.drawable)
                iconView = findViewById(R.id.ad_app_icon)
            }

            // Call to Action
            nativeAd.callToAction?.let {
                findViewById<Button>(R.id.ad_call_to_action).text = it
                callToActionView = findViewById(R.id.ad_call_to_action)
            }

            setNativeAd(nativeAd)
        }

        return nativeAdView
    }
}
